greedy_algo.py

```diff
@@ -72,60 +76,98 @@ def run_replication(self):
   # delta: specify the amount of request increased every time 
   def enough_replica_on_increase(self, delta):
     for c in self.content_set:
-      for a in self.client_set:
-        # add a small amount of requests for content c from client a
-        self.access_map[c][a] += delta
-        # test whether current replicas can handle that much request
-        is_enough = self.enough_replica()
-        # back tracking,
-        self.access_map[c][a] -= delta
-        return is_enough
+      if c in self.access_map:
+        for a in self.access_map[c].keys():
+          # add a small amount of requests for content c from client a
+          self.access_map[c][a] += delta
+          # test whether current replicas can handle that much request
+          is_enough = self.enough_replica()
+          # back tracking,
+          self.access_map[c][a] -= delta
+          if not is_enough:
+            return False
+    return True
 
   def add_replica(self, request_delta, replica_delta):
     I = []
     for c in self.content_set:
-      for a in self.client_set:
-        # add a small amount of requests for content c from client a
-        self.access_map[c][a] += delta
-        # test whether current replicas can handle that much request
-        if not self.enough_replica():
-          I.append((a,c))
-        # back tracking,
-        self.access_map[c][a] -= delta
+      if c in self.access_map:
+        for a in self.access_map[c].keys():
+          # add a small amount of requests for content c from client a
+          self.access_map[c][a] += request_delta
+          # test whether current replicas can handle that much request
+          if not self.enough_replica():
+            I.append((a,c))
+          # back tracking,
+          self.access_map[c][a] -= request_delta
     max_satisfied_num = 0
     best_c = None
     best_s = None
     # find the server s to replicate content c so that  
     # maximum number of starved clients can be satisfied
-    for c in self.content_set:
+    for a, c in I:
       for s in self.server_set:
         satisfied_num = 0
-        for a, I_c in I:
-          if I_c == c:
-            self.access_map[c][a] += request_delta
-            self.replica_map[c][s] += replica_delta
-            if self.enough_replica():
-              satisfied_num += 1
-            self.access_map[c][a] -= request_delta
-            self.replica_map[c][a] -= replica_delta
+        self.access_map[c][a] += request_delta
+        if s not in self.replica_map[c]:
+          self.replica_map[c][s] = 0
+        self.replica_map[c][s] += replica_delta
+        if self.enough_replica():
+          satisfied_num += 1
+        self.access_map[c][a] -= request_delta
+        self.replica_map[c][s] -= replica_delta
+        if self.replica_map[c][s] == 0:
+          self.replica_map[c].pop(s)
         if (satisfied_num > max_satisfied_num):
           max_satisfied_num = satisfied_num
           best_c = c
           best_s = s
```

Comment:
best_s should be the server that maximize satisfied_num. seems this code has logical error though... let me look into it.